neuraxle/metaopt/random.py

```diff
@@ -25,27 +25,264 @@
 """
 
 import copy
-from abc import ABC, abstractmethod
-from typing import List
 import math
+from abc import ABC, abstractmethod
+from typing import List, Callable, Tuple
+
 import numpy as np
 from sklearn.metrics import r2_score
 
-from neuraxle.base import MetaStepMixin, BaseStep
-from neuraxle.steps.numpy import NumpyConcatenateOuterBatch, NumpyConcatenateOnCustomAxis
+from neuraxle.base import MetaStepMixin, BaseStep, ExecutionContext, ExecutionMode, DEFAULT_CACHE_FOLDER
+from neuraxle.data_container import DataContainer
 from neuraxle.steps.loop import StepClonerForEachDataInput
+from neuraxle.steps.numpy import NumpyConcatenateOuterBatch, NumpyConcatenateOnCustomAxis
 
 
-class BaseCrossValidation(MetaStepMixin, BaseStep, ABC):
-    # TODO: assert that set_step was called.
-    # TODO: change default argument of scoring_function...
-    def __init__(self, scoring_function=r2_score, joiner=NumpyConcatenateOuterBatch()):
+class BaseValidation(MetaStepMixin, BaseStep, ABC):
+    """
+    Base class For validation wrappers.
+    It has a scoring function to calculate the score for the validation split.
+
+    .. seealso::
+        :class`ValidationSplitWrapper`,
+        :class`ValidationSplitWrapper`,
+        :class`KFoldCrossValidationWrapper`,
+        :class`AnchoredWalkForwardTimeSeriesCrossValidationWrapper`,
+        :class`WalkForwardTimeSeriesCrossValidationWrapper`
+
+    """
+
+    def __init__(self, scoring_function: Callable = r2_score):
+        """
+        Base class For validation wrappers.
+        It has a scoring function to calculate the score for the validation split.
+
+        :param scoring_function: scoring function with two arguments (y_true, y_pred)
+        :type scoring_function: Callable
+        """
         MetaStepMixin.__init__(self)
         BaseStep.__init__(self)
         self.scoring_function = scoring_function
+
+
+class ValidationSplitWrapper(BaseValidation):
+    """
+    Wrapper for validation split that calculates the score for the validation split.
```

Comment:
We also need to document that the data is NOT shuffled before split, and that if this is something wanted, refer the user to see also the Data Shuffler class, and that the Data Shuffler class can have a seed to allow reproductible research.